netshare/generate/denormalize_fields.py

```diff
@@ -0,0 +1,138 @@
+import csv
+import os
+import random
+import tempfile
+from typing import Dict, List
+
+import numpy as np
+
+from netshare.configs import get_config
+from netshare.generate import generate_api
+from netshare.input_adapters.input_adapter_api import get_canonical_data_dir
+from netshare.learn import learn_api
+from netshare.learn.utils.dataframe_utils import load_dataframe_chunks
+from netshare.utils.field import ContinuousField, Field
+from netshare.utils.logger import logger
+
+
+def _get_fields_names(fields_list: List[Field]) -> List[str]:
+    """
+    This function returns the names of the given fields.
+    """
+    field_names = []
+    for field in fields_list:
+        if isinstance(field.name, list):
+            field_names.extend(field.name)
+        else:
+            field_names.append(field.name)
+    return field_names
+
+
+def _denormalize_by_fields_list(
+    normalized_data: np.ndarray, fields_list: List[Field], is_session_key: bool
+) -> List[np.ndarray]:
+    """
+    This function executes field.denormalize for each of the given field.
+    """
+    denormalized_data = []
+    dim = 0
+
+    canonical_data_dir = get_canonical_data_dir()
+    df, _ = load_dataframe_chunks(canonical_data_dir)
+
+    for field in fields_list:
+        if is_session_key:
+            sub_data = normalized_data[:, dim : dim + field.get_output_dim()]
+        else:
+            sub_data = normalized_data[:, :, dim : dim + field.get_output_dim()]
+
+        sub_data = field.denormalize(sub_data)
+
+        # For session key, if shape looks like (n, ), change it to (n, 1) for consistency
+        if is_session_key and len(sub_data.shape) == 1:
+            sub_data = np.expand_dims(sub_data, axis=1)
+        # For timeseries, if shape looks like (i, j), change it to (i, j, 1) for consistency
+        if not is_session_key and len(sub_data.shape) == 2:
+            sub_data = np.expand_dims(sub_data, axis=2)
+        denormalized_data.append(sub_data)
+        dim += field.get_output_dim()
+    return denormalized_data
+
+
+def write_to_csv(
+    csv_folder: str,
+    session_key_fields: List[Field],
+    timeseries_fields: List[Field],
+    session_key: List[np.ndarray],
+    timeseries: List[np.ndarray],
+    data_gen_flag: np.ndarray,
+    filename: str,
+) -> None:
+    """
+    This function dumps the given data to the given directory as a csv format.
+    `data_gen_flag` is an indicator showing if the time series for this session
+    has ended in this time step.
+    """
+    os.makedirs(csv_folder, exist_ok=True)
+    csv_path = os.path.join(csv_folder, f"data_{filename}_{random.random()}.csv")
```

Comment:
Is there any specific concern to include `random.random()` in the filename?

Reply:
yes. I was afraid that `filename` is not unique enough, thus we will override ourselves.
Checking it retrospectively, I think that this is not necessary, and we can remove it if your prefer